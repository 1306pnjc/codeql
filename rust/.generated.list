ql/lib/codeql/rust/elements/DbFile.qll 056d363e1ba5ec08cacb2e90b8a7a3b47f52ded5dc2289edd4e85921fc50f37e 18e6926f77265a3a6319ca2f3bf3d529d17d46cebdd2ef6753357fdc53c22c70
ql/lib/codeql/rust/elements/DbFileConstructor.qll ea93dc49b23b1c6d800ab9d0b9cacfa9dc661bfdb04b9e6efcad2bdb050fb0ab f7a891b1786604eee57a784733555b677e2580770d51d18073b59e7ca65df1d4
ql/lib/codeql/rust/elements/DbLocation.qll 1f694594e8e4ab65a8781cd443ad4f864447ca88e2cb65504aee5a779393c84d 003ec72275406eb8f5ddd6ccc2b258fb7c906d4bb2c0ef1ba235f291624321ca
ql/lib/codeql/rust/elements/DbLocationConstructor.qll 8848abace985818a5d3a6eddfc4cb200795970146d282b037b4f22ae6230b894 44dba880e17bb1072fa12451ccaae4830fd04dcc61f7403d35510309fde6906e
ql/lib/codeql/rust/elements/Declaration.qll d4ec5c83728f1837243caf2f27d06fd05ecdd2ca440112accff99bfd37b45e5f c1cd9b297be8b69207e75d24b29949b9f71c78406ee0ffd38d0b0810288d6140
ql/lib/codeql/rust/elements/Element.qll 2bffdeb8863a9853d019e56afa268b3d39fb3687785ad636b159656c199997ca 18020582c7d1b89f4073b6deaf793c3072fdc52503a6cc53172c96282f141de4
ql/lib/codeql/rust/elements/File.qll 75144234638417347107b75db80c561379e6372859ebc463276aef716e0ed42b 55bb468873acf24e6828faf7be3ff8963c9a4eb344bfedd95328a51b6f00a226
ql/lib/codeql/rust/elements/Function.qll 220eb1d0e6b49e83b7674a9f505d3f529cf7f4d022ddca44fa6367e0d75daa0f d156cff59ecdcffc0d62041b801025b1aaedbc92d44efd107f93f6fd06c34a6d
ql/lib/codeql/rust/elements/FunctionConstructor.qll a9269b37182c0bf432f9b2b015691da5dbd64819b9bd25445af229d873014a91 69107a7503af14a51e091e6918094a4e9fc316a72de2e1514f001872ce0f2c0c
ql/lib/codeql/rust/elements/Locatable.qll dfc0235cf5aff0ec91d85375ac03998b0e6a69f3722b099330ab0a2161af988a bb1055f59ef7e95e89748765c79f1b5832aa18226f227e689a6801adfc6247ad
ql/lib/codeql/rust/elements/Location.qll 17a222ffe30ecfa252bfd2a5d6ef668f9579824b78d241d576c8c0112a9f93f0 9ae822cd9ce3c70824cccae500651f0cba70ad9c9195b611e7542cb928093b0b
ql/lib/codeql/rust/elements/Module.qll d8995b361cc672f86a314bd53bd3e4d1ddb26b6afde62eb7c380923810785af0 3c10180c812d89a8116ac6e32cbd4d7ac2f549c8a76d327ed75c764b09251d52
ql/lib/codeql/rust/elements/ModuleConstructor.qll 109ed8c1b5c61cc1d3e8613aa8bb8c168dc1943c93b5b622fa79665751b78318 601526c7f56578883d261d14653fdad08329f80fea71de14a5ac5ce671a8d436
ql/lib/codeql/rust/elements/UnknownFile.qll 638ac1c5baef4ab8eb98ef4a37202334290001ca183dab769f729b61f9fc0aa9 fae24f8b29c3e5a124e6c71c9bcb0e2bf4bb033eaf3348fd2db3644ce2803b43
ql/lib/codeql/rust/elements/UnknownLocation.qll eaadf1702e2f5df0c0917dd9e452a3ceb81e5c2d182675c32e988d826ac3fc71 9bf63fbd74c6114678189a62f63795156c0e5b9ad467c020f16842038e84a226
ql/lib/codeql/rust/elements.qll 658d3d26ac2685a077714d41d83ac47f87b0df1c4f0cc7d099c96e06bd1f9d0e 658d3d26ac2685a077714d41d83ac47f87b0df1c4f0cc7d099c96e06bd1f9d0e
ql/lib/codeql/rust/generated/DbFile.qll 4dbf1931124291e0d6a958ae882f8aeef006642f72adc7ff86cffd3a4e9a970a 4dbf1931124291e0d6a958ae882f8aeef006642f72adc7ff86cffd3a4e9a970a
ql/lib/codeql/rust/generated/DbLocation.qll 735d9351b5eb46a3231b528600dddec3a4122c18c210d4d632a8d4ceaf7f02e9 735d9351b5eb46a3231b528600dddec3a4122c18c210d4d632a8d4ceaf7f02e9
ql/lib/codeql/rust/generated/Declaration.qll 4487ac3f5ffa5b92e8628bc04b51e818d4ea1c9a333375cf1b729428d36a4ee7 6481d5e2d99a548f857213a283da75d45db8b3adac949e90fd5d17ceb5e22b54
ql/lib/codeql/rust/generated/Element.qll 21567fa7348dccdf69dd34e73cb6de7cc9c7e0f3f7bb419a1abd787f7dc851a1 21567fa7348dccdf69dd34e73cb6de7cc9c7e0f3f7bb419a1abd787f7dc851a1
ql/lib/codeql/rust/generated/File.qll 2eff5c882d044d2e360fe6382fb55e5a45f6ccb9df323cfa1fae41eae9d2a47f 87e7a906b3a1b4de056952e288ddd7f69fa7cb1bd9dc7dd3972868a9002ac1e4
ql/lib/codeql/rust/generated/Function.qll 6429619a284cadaf64d74ab408f57a399c7e5df2ad182bd38f77b580879d9e33 8c4cfa4f23b5ed461f0f8aa40c5833a953fc9ba3f99584f0f05cf7d7d6a9b617
ql/lib/codeql/rust/generated/Locatable.qll 9e9685bba50ad2220701f3465e63af9331f7f9dc548ad906ff954fc2ce0a4400 78c89b2cc78a357d682ab65310dd474603071f07c1eaaab07711714ce17549f2
ql/lib/codeql/rust/generated/Location.qll bce4c72988ec6fedd1439c60a37c45aa6147c962904709ef9f12206937174be4 d57000571771a2d997c50d9a43ef1c2f075960f847effa0e80ea91fd4c6b4d6c
ql/lib/codeql/rust/generated/Module.qll 2a931a4f2cdb2fee00ed83af045ea63d36b7dbd708e58c30445b5610feaae333 cd62add5c31a509f965aa294f44a1607ec7c62e3a9e3fe9ee063b3c814f4eb62
ql/lib/codeql/rust/generated/ParentChild.qll b44f149bff05a96ee16c88883f06d5d2c159a89423ec32b00765da6964af029e 1d13e54fbc27d9cc29405a21c2f4a63043cbb7aade317184f440a17d3f5645c4
ql/lib/codeql/rust/generated/PureSynthConstructors.qll 5eb1fc4f6a04172c34ae31e4931e4bf1f8b72fbe414c5f644731a45372d13573 5eb1fc4f6a04172c34ae31e4931e4bf1f8b72fbe414c5f644731a45372d13573
ql/lib/codeql/rust/generated/Raw.qll 921cb1afb5c1c3177acb557151755d4f97e7c65f656c5069d6f095b0e078074f a25fdad01e70bbab2d2663b152a4a5844677edcf0a0af2ec93c42dc3248ac9b2
ql/lib/codeql/rust/generated/Synth.qll d278de9c2d06cb7549cd8f2e10ed186827a2ceab6ff46725ca76e78e7fecac72 acacd9afc5ca4a288e037a43375d933c3ba3cd8d08ef122b31695e74be260eb2
ql/lib/codeql/rust/generated/SynthConstructors.qll 35b36df0c4fff05bcbd4ed10b1e6fa2e58fe8d8c818e7805111044825788fc01 35b36df0c4fff05bcbd4ed10b1e6fa2e58fe8d8c818e7805111044825788fc01
ql/lib/codeql/rust/generated/UnknownFile.qll ec9d1a3f15ecbf1743d4e39cb3b2f217aa9b54951c93302c2c4c238c3f0ce595 ec9d1a3f15ecbf1743d4e39cb3b2f217aa9b54951c93302c2c4c238c3f0ce595
ql/lib/codeql/rust/generated/UnknownLocation.qll a19e2838c52d702d268ae530f3dbd6fcd8bb28a237a52636a960f225454103cf a19e2838c52d702d268ae530f3dbd6fcd8bb28a237a52636a960f225454103cf
ql/test/extractor-tests/generated/File/MISSING_SOURCE.txt cc7c395e7c651d62596826b1a0bedf10f35d01b8afeef47600b4ddaf804f406e cc7c395e7c651d62596826b1a0bedf10f35d01b8afeef47600b4ddaf804f406e
ql/test/extractor-tests/generated/Function/MISSING_SOURCE.txt cc7c395e7c651d62596826b1a0bedf10f35d01b8afeef47600b4ddaf804f406e cc7c395e7c651d62596826b1a0bedf10f35d01b8afeef47600b4ddaf804f406e
ql/test/extractor-tests/generated/Module/MISSING_SOURCE.txt cc7c395e7c651d62596826b1a0bedf10f35d01b8afeef47600b4ddaf804f406e cc7c395e7c651d62596826b1a0bedf10f35d01b8afeef47600b4ddaf804f406e
