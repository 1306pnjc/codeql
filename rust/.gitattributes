/.generated.list linguist-generated
/.gitattributes linguist-generated
/ql/lib/codeql/rust/elements/DbFile.qll linguist-generated
/ql/lib/codeql/rust/elements/DbFileConstructor.qll linguist-generated
/ql/lib/codeql/rust/elements/DbLocation.qll linguist-generated
/ql/lib/codeql/rust/elements/DbLocationConstructor.qll linguist-generated
/ql/lib/codeql/rust/elements/Declaration.qll linguist-generated
/ql/lib/codeql/rust/elements/Function.qll linguist-generated
/ql/lib/codeql/rust/elements/FunctionConstructor.qll linguist-generated
/ql/lib/codeql/rust/elements/Locatable.qll linguist-generated
/ql/lib/codeql/rust/elements/Location.qll linguist-generated
/ql/lib/codeql/rust/elements/Module.qll linguist-generated
/ql/lib/codeql/rust/elements/ModuleConstructor.qll linguist-generated
/ql/lib/codeql/rust/elements/UnknownFile.qll linguist-generated
/ql/lib/codeql/rust/elements/UnknownLocation.qll linguist-generated
/ql/lib/codeql/rust/elements.qll linguist-generated
/ql/lib/codeql/rust/generated/DbFile.qll linguist-generated
/ql/lib/codeql/rust/generated/DbLocation.qll linguist-generated
/ql/lib/codeql/rust/generated/Declaration.qll linguist-generated
/ql/lib/codeql/rust/generated/Element.qll linguist-generated
/ql/lib/codeql/rust/generated/File.qll linguist-generated
/ql/lib/codeql/rust/generated/Function.qll linguist-generated
/ql/lib/codeql/rust/generated/Locatable.qll linguist-generated
/ql/lib/codeql/rust/generated/Location.qll linguist-generated
/ql/lib/codeql/rust/generated/Module.qll linguist-generated
/ql/lib/codeql/rust/generated/ParentChild.qll linguist-generated
/ql/lib/codeql/rust/generated/PureSynthConstructors.qll linguist-generated
/ql/lib/codeql/rust/generated/Raw.qll linguist-generated
/ql/lib/codeql/rust/generated/Synth.qll linguist-generated
/ql/lib/codeql/rust/generated/SynthConstructors.qll linguist-generated
/ql/lib/codeql/rust/generated/UnknownFile.qll linguist-generated
/ql/lib/codeql/rust/generated/UnknownLocation.qll linguist-generated
/ql/test/extractor-tests/generated/File/File.ql linguist-generated
/ql/test/extractor-tests/generated/Function/Function.ql linguist-generated
/ql/test/extractor-tests/generated/Module/MISSING_SOURCE.txt linguist-generated
