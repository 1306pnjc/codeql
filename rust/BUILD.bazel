exports_files([
    "codegen.conf",
    "schema.py",
])
