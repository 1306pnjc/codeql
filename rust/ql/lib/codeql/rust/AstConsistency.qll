/**
 * Provides classes for recognizing control flow graph inconsistencies.
 */

private import rust
private import codeql.rust.elements.internal.generated.ParentChild

/**
 * Holds if `e` has more than one `toString()` result.
 */
query predicate multipleToStrings(Element e, string s) {
  s = strictconcat(e.toString(), ", ") and
  strictcount(e.toString()) > 1
}

/**
 * Holds if `e` has more than one `Location`.
 */
query predicate multipleLocations(Locatable e) { strictcount(e.getLocation()) > 1 }

/**
 * Holds if `e` has more than one `getPrimaryQlClasses()` result.
 */
query predicate multiplePrimaryQlClasses(Element e, string s) {
  s = strictconcat(e.getPrimaryQlClasses(), ", ") and
  strictcount(e.getAPrimaryQlClass()) > 1
}

private Element getParent(Element child) { child = getChildAndAccessor(result, _, _) }

/**
 * Holds if `child` has more than one AST parent.
 */
query predicate multipleParents(Element child, Element parent) {
  parent = getParent(child) and
  strictcount(getParent(child)) > 1
}
