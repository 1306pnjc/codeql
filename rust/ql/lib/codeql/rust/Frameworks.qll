/**
 * This file imports all models of frameworks and libraries.
 */
