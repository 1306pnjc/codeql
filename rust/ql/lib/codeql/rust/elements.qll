// generated by codegen
/**
 * This module exports all modules providing `Element` subclasses.
 */

import codeql.rust.elements.DbFile
import codeql.rust.elements.DbLocation
import codeql.rust.elements.Declaration
import codeql.rust.elements.Element
import codeql.rust.elements.File
import codeql.rust.elements.Function
import codeql.rust.elements.Locatable
import codeql.rust.elements.Location
import codeql.rust.elements.Module
import codeql.rust.elements.UnknownFile
import codeql.rust.elements.UnknownLocation
