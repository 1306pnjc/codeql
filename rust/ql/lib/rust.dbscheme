// generated by codegen

// from prefix.dbscheme
/**
 * The source location of the snapshot.
 */
sourceLocationPrefix(
  string prefix: string ref
);


// from schema.py

@element =
  @file
| @locatable
| @location
| @unimplemented
;

@file =
  @db_file
;

#keyset[id]
files(
  int id: @file ref,
  string name: string ref
);

@locatable =
  @ast_node
;

#keyset[id]
locatable_locations(
  int id: @locatable ref,
  int location: @location ref
);

@location =
  @db_location
;

#keyset[id]
locations(
  int id: @location ref,
  int file: @file ref,
  int start_line: int ref,
  int start_column: int ref,
  int end_line: int ref,
  int end_column: int ref
);

@unimplemented =
  @generic_arg_list
| @path
| @type
| @unimplemented_declaration
;

@ast_node =
  @declaration
| @expr
| @generic_arg_list
| @label
| @match_arm
| @pat
| @path
| @record_expr_field
| @record_pat_field
| @stmt
| @type
;

db_files(
  unique int id: @db_file
);

db_locations(
  unique int id: @db_location
);

@declaration =
  @function
| @module
| @unimplemented_declaration
;

@expr =
  @array_expr
| @asm_expr
| @await_expr
| @become_expr
| @bin_expr
| @block_expr_base
| @box_expr
| @break_expr
| @call_expr
| @cast_expr
| @closure_expr
| @const_expr
| @continue_expr
| @field_expr
| @if_expr
| @index_expr
| @let_expr
| @literal
| @loop_expr
| @match_expr
| @method_call_expr
| @missing_expr
| @offset_of_expr
| @path_expr
| @prefix_expr
| @range_expr
| @record_expr
| @ref_expr
| @return_expr
| @tuple_expr
| @underscore_expr
| @yeet_expr
| @yield_expr
;

generic_arg_lists(
  unique int id: @generic_arg_list
);

labels(
  unique int id: @label,
  string name: string ref
);

match_arms(
  unique int id: @match_arm,
  int pat: @pat ref,
  int expr: @expr ref
);

#keyset[id]
match_arm_guards(
  int id: @match_arm ref,
  int guard: @expr ref
);

@pat =
  @box_pat
| @const_block_pat
| @ident_pat
| @literal_pat
| @missing_pat
| @or_pat
| @path_pat
| @range_pat
| @record_pat
| @ref_pat
| @slice_pat
| @tuple_pat
| @tuple_struct_pat
| @wildcard_pat
;

paths(
  unique int id: @path
);

record_expr_fields(
  unique int id: @record_expr_field,
  string name: string ref,
  int expr: @expr ref
);

record_pat_fields(
  unique int id: @record_pat_field,
  string name: string ref,
  int pat: @pat ref
);

@stmt =
  @expr_stmt
| @item_stmt
| @let_stmt
;

types(
  unique int id: @type
);

@array_expr =
  @element_list_expr
| @repeat_expr
;

asm_exprs(
  unique int id: @asm_expr,
  int expr: @expr ref
);

await_exprs(
  unique int id: @await_expr,
  int expr: @expr ref
);

become_exprs(
  unique int id: @become_expr,
  int expr: @expr ref
);

bin_exprs(
  unique int id: @bin_expr,
  int lhs: @expr ref,
  int rhs: @expr ref
);

#keyset[id]
bin_expr_ops(
  int id: @bin_expr ref,
  string op: string ref
);

@block_expr_base =
  @async_block_expr
| @block_expr
| @unsafe_block_expr
;

#keyset[id, index]
block_expr_base_statements(
  int id: @block_expr_base ref,
  int index: int ref,
  int statement: @stmt ref
);

#keyset[id]
block_expr_base_tails(
  int id: @block_expr_base ref,
  int tail: @expr ref
);

box_exprs(
  unique int id: @box_expr,
  int expr: @expr ref
);

box_pats(
  unique int id: @box_pat,
  int inner: @pat ref
);

break_exprs(
  unique int id: @break_expr
);

#keyset[id]
break_expr_exprs(
  int id: @break_expr ref,
  int expr: @expr ref
);

#keyset[id]
break_expr_labels(
  int id: @break_expr ref,
  int label: @label ref
);

call_exprs(
  unique int id: @call_expr,
  int callee: @expr ref
);

#keyset[id, index]
call_expr_args(
  int id: @call_expr ref,
  int index: int ref,
  int arg: @expr ref
);

#keyset[id]
call_expr_is_assignee_expr(
  int id: @call_expr ref
);

cast_exprs(
  unique int id: @cast_expr,
  int expr: @expr ref,
  int type_: @type ref
);

closure_exprs(
  unique int id: @closure_expr,
  int body: @expr ref,
  string closure_kind: string ref
);

#keyset[id, index]
closure_expr_args(
  int id: @closure_expr ref,
  int index: int ref,
  int arg: @pat ref
);

#keyset[id, index]
closure_expr_arg_types(
  int id: @closure_expr ref,
  int index: int ref,
  int arg_type: @type ref
);

#keyset[id]
closure_expr_ret_types(
  int id: @closure_expr ref,
  int ret_type: @type ref
);

#keyset[id]
closure_expr_is_move(
  int id: @closure_expr ref
);

const_block_pats(
  unique int id: @const_block_pat,
  int expr: @expr ref
);

const_exprs(
  unique int id: @const_expr,
  int expr: @expr ref
);

continue_exprs(
  unique int id: @continue_expr
);

#keyset[id]
continue_expr_labels(
  int id: @continue_expr ref,
  int label: @label ref
);

expr_stmts(
  unique int id: @expr_stmt,
  int expr: @expr ref
);

#keyset[id]
expr_stmt_has_semicolon(
  int id: @expr_stmt ref
);

field_exprs(
  unique int id: @field_expr,
  int expr: @expr ref,
  string name: string ref
);

functions(
  unique int id: @function,
  string name: string ref,
  int body: @expr ref
);

ident_pats(
  unique int id: @ident_pat,
  string binding_id: string ref
);

#keyset[id]
ident_pat_subpats(
  int id: @ident_pat ref,
  int subpat: @pat ref
);

if_exprs(
  unique int id: @if_expr,
  int condition: @expr ref,
  int then: @expr ref
);

#keyset[id]
if_expr_elses(
  int id: @if_expr ref,
  int else: @expr ref
);

index_exprs(
  unique int id: @index_expr,
  int base: @expr ref,
  int index: @expr ref
);

#keyset[id]
index_expr_is_assignee_expr(
  int id: @index_expr ref
);

item_stmts(
  unique int id: @item_stmt
);

let_exprs(
  unique int id: @let_expr,
  int pat: @pat ref,
  int expr: @expr ref
);

let_stmts(
  unique int id: @let_stmt,
  int pat: @pat ref
);

#keyset[id]
let_stmt_types(
  int id: @let_stmt ref,
  int type_: @type ref
);

#keyset[id]
let_stmt_initializers(
  int id: @let_stmt ref,
  int initializer: @expr ref
);

#keyset[id]
let_stmt_elses(
  int id: @let_stmt ref,
  int else: @expr ref
);

literals(
  unique int id: @literal
);

literal_pats(
  unique int id: @literal_pat,
  int expr: @expr ref
);

loop_exprs(
  unique int id: @loop_expr,
  int body: @expr ref
);

#keyset[id]
loop_expr_labels(
  int id: @loop_expr ref,
  int label: @label ref
);

match_exprs(
  unique int id: @match_expr,
  int expr: @expr ref
);

#keyset[id, index]
match_expr_branches(
  int id: @match_expr ref,
  int index: int ref,
  int branch: @match_arm ref
);

method_call_exprs(
  unique int id: @method_call_expr,
  int receiver: @expr ref,
  string method_name: string ref
);

#keyset[id, index]
method_call_expr_args(
  int id: @method_call_expr ref,
  int index: int ref,
  int arg: @expr ref
);

#keyset[id]
method_call_expr_generic_args(
  int id: @method_call_expr ref,
  int generic_args: @generic_arg_list ref
);

missing_exprs(
  unique int id: @missing_expr
);

missing_pats(
  unique int id: @missing_pat
);

modules(
  unique int id: @module
);

#keyset[id, index]
module_declarations(
  int id: @module ref,
  int index: int ref,
  int declaration: @declaration ref
);

offset_of_exprs(
  unique int id: @offset_of_expr,
  int container: @type ref
);

#keyset[id, index]
offset_of_expr_fields(
  int id: @offset_of_expr ref,
  int index: int ref,
  string field: string ref
);

or_pats(
  unique int id: @or_pat
);

#keyset[id, index]
or_pat_args(
  int id: @or_pat ref,
  int index: int ref,
  int arg: @pat ref
);

path_exprs(
  unique int id: @path_expr,
  int path: @path ref
);

path_pats(
  unique int id: @path_pat,
  int path: @path ref
);

prefix_exprs(
  unique int id: @prefix_expr,
  int expr: @expr ref,
  string op: string ref
);

range_exprs(
  unique int id: @range_expr
);

#keyset[id]
range_expr_lhs(
  int id: @range_expr ref,
  int lhs: @expr ref
);

#keyset[id]
range_expr_rhs(
  int id: @range_expr ref,
  int rhs: @expr ref
);

#keyset[id]
range_expr_is_inclusive(
  int id: @range_expr ref
);

range_pats(
  unique int id: @range_pat
);

#keyset[id]
range_pat_starts(
  int id: @range_pat ref,
  int start: @pat ref
);

#keyset[id]
range_pat_ends(
  int id: @range_pat ref,
  int end: @pat ref
);

record_exprs(
  unique int id: @record_expr
);

#keyset[id]
record_expr_paths(
  int id: @record_expr ref,
  int path: @path ref
);

#keyset[id, index]
record_expr_fields(
  int id: @record_expr ref,
  int index: int ref,
  int field: @record_expr_field ref
);

#keyset[id]
record_expr_spreads(
  int id: @record_expr ref,
  int spread: @expr ref
);

#keyset[id]
record_expr_has_ellipsis(
  int id: @record_expr ref
);

#keyset[id]
record_expr_is_assignee_expr(
  int id: @record_expr ref
);

record_pats(
  unique int id: @record_pat
);

#keyset[id]
record_pat_paths(
  int id: @record_pat ref,
  int path: @path ref
);

#keyset[id, index]
record_pat_args(
  int id: @record_pat ref,
  int index: int ref,
  int arg: @record_pat_field ref
);

#keyset[id]
record_pat_has_ellipsis(
  int id: @record_pat ref
);

ref_exprs(
  unique int id: @ref_expr,
  int expr: @expr ref
);

#keyset[id]
ref_expr_is_raw(
  int id: @ref_expr ref
);

#keyset[id]
ref_expr_is_mut(
  int id: @ref_expr ref
);

ref_pats(
  unique int id: @ref_pat,
  int pat: @pat ref
);

#keyset[id]
ref_pat_is_mut(
  int id: @ref_pat ref
);

return_exprs(
  unique int id: @return_expr
);

#keyset[id]
return_expr_exprs(
  int id: @return_expr ref,
  int expr: @expr ref
);

slice_pats(
  unique int id: @slice_pat
);

#keyset[id, index]
slice_pat_prefixes(
  int id: @slice_pat ref,
  int index: int ref,
  int prefix: @pat ref
);

#keyset[id]
slice_pat_slice(
  int id: @slice_pat ref,
  int slice: @pat ref
);

#keyset[id, index]
slice_pat_suffixes(
  int id: @slice_pat ref,
  int index: int ref,
  int suffix: @pat ref
);

tuple_exprs(
  unique int id: @tuple_expr
);

#keyset[id, index]
tuple_expr_exprs(
  int id: @tuple_expr ref,
  int index: int ref,
  int expr: @expr ref
);

#keyset[id]
tuple_expr_is_assignee_expr(
  int id: @tuple_expr ref
);

tuple_pats(
  unique int id: @tuple_pat
);

#keyset[id, index]
tuple_pat_args(
  int id: @tuple_pat ref,
  int index: int ref,
  int arg: @pat ref
);

#keyset[id]
tuple_pat_ellipsis_indices(
  int id: @tuple_pat ref,
  int ellipsis_index: int ref
);

tuple_struct_pats(
  unique int id: @tuple_struct_pat
);

#keyset[id]
tuple_struct_pat_paths(
  int id: @tuple_struct_pat ref,
  int path: @path ref
);

#keyset[id, index]
tuple_struct_pat_args(
  int id: @tuple_struct_pat ref,
  int index: int ref,
  int arg: @pat ref
);

#keyset[id]
tuple_struct_pat_ellipsis_indices(
  int id: @tuple_struct_pat ref,
  int ellipsis_index: int ref
);

underscore_exprs(
  unique int id: @underscore_expr
);

unimplemented_declarations(
  unique int id: @unimplemented_declaration
);

wildcard_pats(
  unique int id: @wildcard_pat
);

yeet_exprs(
  unique int id: @yeet_expr
);

#keyset[id]
yeet_expr_exprs(
  int id: @yeet_expr ref,
  int expr: @expr ref
);

yield_exprs(
  unique int id: @yield_expr
);

#keyset[id]
yield_expr_exprs(
  int id: @yield_expr ref,
  int expr: @expr ref
);

async_block_exprs(
  unique int id: @async_block_expr
);

block_exprs(
  unique int id: @block_expr
);

#keyset[id]
block_expr_labels(
  int id: @block_expr ref,
  int label: @label ref
);

element_list_exprs(
  unique int id: @element_list_expr
);

#keyset[id, index]
element_list_expr_elements(
  int id: @element_list_expr ref,
  int index: int ref,
  int element: @expr ref
);

#keyset[id]
element_list_expr_is_assignee_expr(
  int id: @element_list_expr ref
);

repeat_exprs(
  unique int id: @repeat_expr,
  int initializer: @expr ref,
  int repeat: @expr ref
);

unsafe_block_exprs(
  unique int id: @unsafe_block_expr
);
