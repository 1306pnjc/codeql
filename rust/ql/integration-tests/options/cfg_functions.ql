import rust

from Function f
where f.getLocation().getFile().getBaseName() = "cfg.rs" and f.hasExtendedCanonicalPath()
select f
