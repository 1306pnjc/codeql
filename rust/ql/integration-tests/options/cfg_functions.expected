| src/cfg.rs:7:1:8:19 | cfg_no_flag |
| src/cfg.rs:10:1:11:18 | cfg_no_key |
